pandas/core/internals/blocks.py

```diff
@@ -781,16 +788,15 @@ def replace(
         inplace = validate_bool_kwarg(inplace, "inplace")
         original_to_replace = to_replace
 
-        # try to replace, if we raise an error, convert to ObjectBlock and
+        # If we cannot replace with own dtype, convert to ObjectBlock and
         # retry
-        values = self._coerce_values(self.values)
-        try:
-            to_replace = self._try_coerce_args(to_replace)
-        except (TypeError, ValueError):
+        if not self._can_hold_element(to_replace):
+            # TODO: we should be able to infer at this point that there is
+            #  nothing to replace
             # GH 22083, TypeError or ValueError occurred within error handling
             # causes infinite loop. Cast and retry only if not objectblock.
             if is_object_dtype(self):
-                raise
+                raise AssertionError
```

Comment:
this makes this block *different* than all others, why?

Reply:
in the status quo, `raise` re-raises the exception that is caught on 789.  But now we don't have that to re-raise, so we need to raise a new exception.  The `raise` on 793 is not reached (and should not be reachable)

Reply:
why don't we instead define an internal exception, say DtypeInvalidException (or better name) which inherits this then for flow control.